### PR TITLE
Add docs for CoderGrammar

### PR DESCRIPTION
**Pull request by anish749**
I was trying to speed up my compilation and resorted to adding implicit definitions explicitly using `Coder.gen`, `Coder.kryo` and `Coder.xmap` which means I am somewhat relying on this api.

I am adding a few docs, so that someone reading our code can debug / understand what these functions are doing.

### CODE REVIEW

File: scio-core/src/main/scala/com/spotify/scio/coders/DerivedCoder.scala [hunk 1]
```diff
@@ -41,7 +41,7 @@ private object Derived extends Serializable {
       i = i + 1
     }
 
-    val stack = CoderException.prepareStackTrace
+    lazy val stack = CoderException.prepareStackTrace
```
Comment (jto) [inline scio-core/src/main/scala/com/spotify/scio/coders/DerivedCoder.scala]: Why did you change that ?

Comment (anish749) [reply scio-core/src/main/scala/com/spotify/scio/coders/DerivedCoder.scala]: I was trying to delay the stack trace preparation to see if the speed improves. its negligible, shall i remove this?

Comment (jto) [reply scio-core/src/main/scala/com/spotify/scio/coders/DerivedCoder.scala]: Yes. The purpose is to pre-generate the stack trace to display the call site instead of beam's internal when an exception occurs. Maky it lazy is a bug.

Comment (anish749) [reply scio-core/src/main/scala/com/spotify/scio/coders/DerivedCoder.scala]: I see now. Thanks for explaining.

**Comment by anish749** [comment]
Thanks for the reviews..
I was wondering if the information in this PR is correct. I kind of guessed my way through so not very sure if everything is correct and if they advise best practices.

**Comment by jto** [comment]
Everything is correct. Thanks for the PR @anish749 :)
